parlai/core/worlds.py

```diff
@@ -975,6 +975,209 @@ def shutdown(self):
+            # anything with the same cost for increased stochasticity,
+            # while not really wasting much padding.
+            return tuple(
+                self._ceil(len(obs[key]))
+                for key in ['text_vec', 'labels_vec', 'eval_labels_vec']
+                if key in obs
+            )
+        else:
+            return None
+
+    def parley(self):
+        # first make sure that all the worlds are processed in the queue
+        indices = []
+        for i in range(self._BUFFER_SIZE):
+            if self._scores[i] is not None:
+                indices.append(i)
+                continue
+            if self.worlds[i].epoch_done():
+                continue
+
+            if hasattr(self.world, 'parley_init'):
+                self.worlds[i].parley_init()
+
+            act = self.worlds[i].get_agents()[0].act()
+            obs = self.worlds[i].get_agents()[1].observe(act)
+            self._obs[i] = obs
+
+            self._scores[i] = self._score(obs)
+            if self._scores[i] is not None:
+                indices.append(i)
+
+        # quick invariant checks
+        assert len(indices) != 0, "DynamicBatchWorld ran out of data!"
+        assert not any(self._scores[i] is None for i in indices)
+
+        # sort all the indices by their score, so that we can find similarly lengthed
+        # items in O(1)
+        indices = sorted(indices, key=lambda i: self._scores[i] + (self.rng.random(),))
+
+        # now let's build the batch
+        batch = []
+
+        # start with a random item. index_index is the lookup into indices, but
+        # index is the actual world.
+        width = 0
+        index_index = random.randint(0, len(indices) - 1)
+
+        # we picked a random spot, but we can get better packing if we start at
+        # the last example with the same score, since we always pick move down
+        # to smaller examples.
+        while index_index < len(indices) - 1 and (
+            sum(self._scores[indices[index_index]])
+            == sum(self._scores[indices[index_index + 1]])
+        ):
+            index_index += 1
+
+        # quit early if we eat our full buffer
+        while indices:
+            index = indices[index_index]
+            this_width = self._ceil(sum(self._scores[index]))
+            new_width = max(width, this_width)
+            # compute the cost of the new batch
+            new_size = new_width * self._ceil(len(batch) + 1)
+            if new_size <= self.max_size:
+                # cool, this one fits, let's add it
+                width = new_width
+                batch.append(index)
+                indices.pop(index_index)
+                index_index = max(index_index - 1, 0)
+            else:
+                # we'd overfill our buffer, give up
+                break
+
+        # Always have a batch size that's a multiple of 4, for fp16's sake.
+        while len(batch) > 4 and len(batch) % 4 != 0:
+            batch.pop(0)
```

Comment:
why pop the beginning? is this getting rid of the largest examples?

Reply:
Yeah, I don't think it matters what you pop really. Just trying to round it down to a multiple of 4. We could remove the shortest (they are easier to pack later...)

Reply:
That was my thought too, that it'd be easier to pack those in later (i.e. let's get rid of the longer episodes first)